### PR TITLE
V2 create project - token form

### PR DESCRIPTION
**Pull request by johnnyd-eth**
## What does this PR do and why?

Towards https://github.com/jbx-protocol/juice-interface/issues/543

This one is stable and ready to go, saving to and loading from Redux nicely.

Only thing I didn't deal with as far as I could see is typing redemption rate straight into the number input being a bit glitchy - puts a extra '%' in there and looks a bit dodge

## Screenshots or screen recordings

_If applicable, provide screenshots or screen recordings to demonstrate the
changes._

## Acceptance checklist

- [ ] I have evaluated the [Approval Guidelines](../CONTRIBUTING.md#approval-guidelines) for this PR.
- [ ] I have tested this PR in [all supported browsers](../CONTRIBUTING.md#browser-support).

### CODE REVIEW

**Comment by johnnyd-eth** [comment]
I think this one is good for review @aeolianeth. Only thing I could see incomplete is the comment I made in the description. LMK if you find some others

File: src/components/shared/formItems/BondingCurve/ProjectBondingCurveRate.tsx [hunk 1]
```diff
@@ -112,13 +117,21 @@ export default function ProjectBondingCurveRate({
       label={
         hideLabel ? undefined : (
           <div>
-            <Trans>Bonding curve rate</Trans>{' '}
+            <span>{label ? label : t`Bonding curve rate`}</span>{' '}
```
Comment (aeolianeth) [inline src/components/shared/formItems/BondingCurve/ProjectBondingCurveRate.tsx:120]: we can basically avoid `t` in JSX altogether for the most part. no big deal though

File: src/components/shared/formItems/BondingCurve/ProjectBondingCurveRate.tsx [hunk 2]
```diff
@@ -112,13 +117,21 @@ export default function ProjectBondingCurveRate({
...
             {toggleDisabled ? (
-              <Switch checked={!disabled} onChange={toggleDisabled} />
+              <React.Fragment>
```
Comment (aeolianeth) [inline src/components/shared/formItems/BondingCurve/ProjectBondingCurveRate.tsx:122]: note shorthand for this is `<>`

File: src/components/v2/V2Create/tabs/TokenTabContent.tsx [hunk 3]
```diff
@@ -1,39 +1,235 @@
...
+          <Trans>
+            <strong>Note: </strong>Updates you make to token attributes will{' '}
+            <i>not be applied immediately</i> and only take effect in{' '}
+            <i>upcoming funding cycles.</i>
+          </Trans>
+        </p>
+        <Form form={tokenForm} layout="vertical">
+          <FormItems.ProjectReserved
+            value={tokenForm.getFieldValue('reservedRate') ?? reduxReservedRate}
+            name="reservedRate"
```
Comment (aeolianeth) [inline src/components/v2/V2Create/tabs/TokenTabContent.tsx:129]: pls remove `name` whereever we have an `onChange` as the antd form binding can cause unexpected behavior.

File: src/components/v2/V2Create/tabs/TokenTabContent.tsx [hunk 4]
```diff
@@ -1,39 +1,235 @@
...
+                  }
+                : undefined
+            }
+          />
+          <Form.Item>
+            <Button
+              htmlType="submit"
+              type="primary"
+              onClick={async () => {
+                await tokenForm.validateFields()
```
Comment (aeolianeth) [inline src/components/v2/V2Create/tabs/TokenTabContent.tsx:225]: onFinish

File: src/utils/formatNumber.ts [hunk 5]
```diff
@@ -57,6 +57,12 @@ export const parsePermille = (amt?: string | number) =>
 export const fromPermille = (amt?: BigNumberish) =>
   amt ? (BigNumber.from(amt).toNumber() / 10).toString() : '0'
 
+// // Percent: x/1000
+// export const parsePercent = (amt?: string | number) =>
```
Comment (aeolianeth) [inline src/utils/formatNumber.ts:61]: can we delete?

File: src/utils/fundingCycle.ts [hunk 6]
```diff
@@ -85,7 +86,10 @@ export const hasFundingTarget = (
 ) => fundingCycle.target.lt(constants.MaxUint256)
 
 export const hasFundingDuration = (
-  fundingCycle: Pick<V1FundingCycle | EditingV1FundingCycle, 'duration'>,
+  fundingCycle: Pick<
+    V1FundingCycle | EditingV1FundingCycle | V2FundingCycleData,
```
Comment (aeolianeth) [inline src/utils/fundingCycle.ts:90]: **issue**; lets make a new v2 file for this. I don't think we should be mixing v1 and v2 utils.

Don't worry about moving this file to utils/v1/ as I'll handle it in https://github.com/jbx-protocol/juice-interface/pull/569

File: src/utils/stripPercent.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// Strips '%'s from strings. Used in form items such as reserved rate and discount
+export default function stripPercent(val: any) {
```
Comment (aeolianeth) [inline src/utils/stripPercent.ts:2]: **question**; unused?

File: src/utils/v2/serializers.ts [hunk 8]
```diff
@@ -40,8 +42,10 @@ export const serializeV2FundingCycleMetadata = (
   fundingCycleMetadata: V2FundingCycleMetadata,
 ): SerializedV2FundingCycleMetadata => ({
   reservedRate: fromPermille(fundingCycleMetadata.reservedRate),
-  redemptionRate: fromWad(fundingCycleMetadata.redemptionRate),
-  ballotRedemptionRate: fromWad(fundingCycleMetadata.ballotRedemptionRate),
+  redemptionRate: fromPerbicent(fundingCycleMetadata.redemptionRate),
```
Comment (aeolianeth) [inline src/utils/v2/serializers.ts:45]: **praise** thanks for updating these

File: src/components/v2/V2Create/tabs/TokenTabContent.tsx [hunk 9]
```diff
@@ -1,39 +1,235 @@
...
   useEffect(() => {
     resetTokenForm()
   }, [resetTokenForm])
 
+  // Assume the first item is the one of interest.
+  const fundAccessConstraint = fundAccessConstraints[0] as
+    | SerializedV2FundAccessConstraint
+    | undefined
+
+  const _hasFundingTarget = Boolean(
```
Comment (aeolianeth) [inline src/components/v2/V2Create/tabs/TokenTabContent.tsx:109]: **suggestion** can we avoid underscore in this name?

File: src/components/v2/V2Create/tabs/TokenTabContent.tsx [hunk 10]
```diff
@@ -1,39 +1,235 @@
...
+    parseFloat(fundAccessConstraint?.distributionLimit ?? '0'),
+  )
+  const _hasFundingDuration = Boolean(
```
Comment (aeolianeth) [inline src/components/v2/V2Create/tabs/TokenTabContent.tsx:112]: **praise** nice use of `Boolean`

**Review by aeolianeth — COMMENTED** [review]
@johnnyd-eth awesome work once more, a few comments

File: src/components/v2/V2Create/tabs/TokenTabContent.tsx (hunk 3 above)
Comment (johnnyd-eth) [reply src/components/v2/V2Create/tabs/TokenTabContent.tsx:129]: Got it

File: src/utils/fundingCycle.ts (hunk 6 above)
Comment (johnnyd-eth) [reply src/utils/fundingCycle.ts:90]: @aeolianeth yep moved into separate file `fundingCycleV2.ts`

File: src/components/v2/V2Create/tabs/TokenTabContent.tsx (hunk 9 above)
Comment (johnnyd-eth) [reply src/components/v2/V2Create/tabs/TokenTabContent.tsx:109]: Removed these and used the new functions in `utils/fundingCycleV2`

Comment (aeolianeth) [reply src/components/v2/V2Create/tabs/TokenTabContent.tsx]: respecc

File: src/components/shared/ReconfigurationDrawer.tsx [hunk 11]
```diff
@@ -90,6 +90,7 @@ export default function ReconfigurationDrawer({
         {ballotStrategies().map((s: Strategy, i) => (
           <ReconfigurationOption
             title={s.name}
+            key={i}
```
Comment (aeolianeth) [inline src/components/shared/ReconfigurationDrawer.tsx:93]: NICE

**Comment by aeolianeth** [comment]
Fresh, merging 🚀

**Pull request merged**